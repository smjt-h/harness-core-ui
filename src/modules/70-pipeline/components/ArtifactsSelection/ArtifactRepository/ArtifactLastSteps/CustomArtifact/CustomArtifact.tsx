/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import {
  Formik,
  Layout,
  Button,
  StepProps,
  Text,
  ButtonVariation,
  FormInput,
  getMultiTypeFromValue,
  MultiTypeInputType
} from '@harness/uicore'
import { Form } from 'formik'
import * as Yup from 'yup'
import { FontVariation } from '@harness/design-system'
import { merge } from 'lodash-es'
import { useStrings } from 'framework/strings'

import type {
  ArtifactType,
  ImagePathProps,
  CustomArtifactSource
} from '@pipeline/components/ArtifactsSelection/ArtifactInterface'
import type { ConnectorConfigDTO } from 'services/cd-ng'
import { ConfigureOptions } from '@common/components/ConfigureOptions/ConfigureOptions'
import { getArtifactFormData } from '@pipeline/components/ArtifactsSelection/ArtifactUtils'
import { ArtifactIdentifierValidation, ModalViewFor } from '../../../ArtifactHelper'
import SideCarArtifactIdentifier from '../SideCarArtifactIdentifier'
import css from '../../ArtifactConnector.module.scss'

export function CustomArtifact({
  context,
  handleSubmit,
  expressions,
  allowableTypes,
  prevStepData,
  initialValues,
  previousStep,
  artifactIdentifiers,
  isReadonly = false,
  selectedArtifact
}: StepProps<ConnectorConfigDTO> & ImagePathProps): React.ReactElement {
  const { getString } = useStrings()

  const schemaObject = {
    version: Yup.string().trim().required(getString('validation.nexusVersion'))
  }

  const primarySchema = Yup.object().shape(schemaObject)
  const sidecarSchema = Yup.object().shape({
    ...schemaObject,
    ...ArtifactIdentifierValidation(
      artifactIdentifiers,
      initialValues?.identifier,
      getString('pipeline.uniqueIdentifier')
    )
  })

  const getInitialValues = (): CustomArtifactSource => {
    return getArtifactFormData(
      initialValues,
      selectedArtifact as ArtifactType,
      context === ModalViewFor.SIDECAR
    ) as CustomArtifactSource
  }
  const submitFormData = (formData: CustomArtifactSource & { connectorId?: string }): void => {
    const artifactObj = {
      spec: {
        version: formData?.version
      }
    }
    if (context === ModalViewFor.SIDECAR) {
      merge(artifactObj, { identifier: formData?.identifier })
    }
    handleSubmit(artifactObj)
  }

  return (
    <Layout.Vertical spacing="medium" className={css.firstep}>
      <Text font={{ variation: FontVariation.H3 }} margin={{ bottom: 'medium' }}>
        {getString('pipeline.artifactsSelection.artifactDetails')}
      </Text>
      <Formik
        initialValues={getInitialValues()}
        formName="imagePath"
        validationSchema={context === ModalViewFor.SIDECAR ? sidecarSchema : primarySchema}
        onSubmit={formData => {
          submitFormData({
            ...prevStepData,
            ...formData
          })
        }}
      >
        {formik => (
          <Form>
            <div className={css.connectorForm}>
              {context === ModalViewFor.SIDECAR && <SideCarArtifactIdentifier />}

              <div className={css.imagePathContainer}>
                <FormInput.MultiTextInput
                  label={getString('version')}
                  name="version"
                  placeholder={getString('pipeline.artifactsSelection.versionPlaceholder')}
                  multiTextInputProps={{
                    expressions,
                    allowableTypes
                  }}
                />

                {getMultiTypeFromValue(formik.values.version) === MultiTypeInputType.RUNTIME && (
                  <div className={css.configureOptions}>
                    <ConfigureOptions
                      style={{ alignSelf: 'center' }}
                      value={formik.values?.version as string}
                      type={getString('string')}
                      variableName="version"
                      showRequiredField={false}
                      showDefaultField={false}
                      showAdvanced={true}
                      onChange={value => {
                        formik.setFieldValue('version', value)
                      }}
                      isReadonly={isReadonly}
                    />
                  </div>
                )}
              </div>
            </div>
            <Layout.Horizontal spacing="medium">
              <Button
                variation={ButtonVariation.SECONDARY}
                text={getString('back')}
                icon="chevron-left"
                onClick={() => previousStep?.(prevStepData)}
              />
              <Button
                variation={ButtonVariation.PRIMARY}
                type="submit"
                text={getString('submit')}
                rightIcon="chevron-right"
              />
            </Layout.Horizontal>
          </Form>
        )}
      </Formik>
    </Layout.Vertical>
  )
}
