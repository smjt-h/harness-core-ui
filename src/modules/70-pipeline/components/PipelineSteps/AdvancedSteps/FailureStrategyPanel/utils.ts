/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { SetStateAction } from 'react'
import type { FormikErrors, FormikProps } from 'formik'
import { isEmpty } from 'lodash-es'
import { Intent } from '@blueprintjs/core'

import type {
  RetryFailureActionConfig,
  IgnoreFailureActionConfig,
  ManualInterventionFailureActionConfig,
  AbortFailureActionConfig,
  StepGroupFailureActionConfig,
  MarkAsSuccessFailureActionConfig,
  StageRollbackFailureActionConfig,
  FailureStrategyConfig,
  OnFailureConfig
} from 'services/cd-ng'

export type AllActions =
  | RetryFailureActionConfig
  | IgnoreFailureActionConfig
  | ManualInterventionFailureActionConfig
  | AbortFailureActionConfig
  | StepGroupFailureActionConfig
  | MarkAsSuccessFailureActionConfig
  | StageRollbackFailureActionConfig

export interface AllFailureStrategyConfig extends FailureStrategyConfig {
  onFailure: OnFailureConfig & { action: AllActions }
}

export function hasItems<T>(data?: T[]): boolean {
  return Array.isArray(data) && data.length > 0
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function findTabWithErrors(errors?: FormikErrors<any>): number {
  const fsErrors = errors?.failureStrategies

  return Array.isArray(fsErrors)
    ? // eslint-disable-next-line @typescript-eslint/no-explicit-any
      fsErrors.findIndex((err: any) => !isEmpty(err))
    : -1
}

export interface FormState {
  failureStrategies?: AllFailureStrategyConfig[]
}

export interface HandleChangeInStrategiesProps {
  formValues: FormState
  selectedStrategyNum: number
  setSelectedStrategyNum: (value: SetStateAction<number>) => void
  setFormikState: FormikProps<FormState>['setFormikState']
}

export function handleChangeInStrategies({
  formValues,
  selectedStrategyNum,
  setSelectedStrategyNum,
  setFormikState
}: HandleChangeInStrategiesProps): void {
  const fs = formValues.failureStrategies
  /* istanbul ignore else */
  if (Array.isArray(fs)) {
    /* istanbul ignore else */
    if (selectedStrategyNum >= fs.length) {
      // select the new last tab, if the last tab was deleted
      setSelectedStrategyNum(Math.max(0, fs.length - 1))
    }

    /* istanbul ignore else */
    if (fs.length === 0) {
      // reset errors when all the tabs are deleted
      setFormikState({ errors: {}, submitCount: 0 })
    }
  }
}

export function getTabIntent(i: number, selectedStrategyNum: number): Intent {
  return i === selectedStrategyNum ? Intent.PRIMARY : Intent.NONE
}
