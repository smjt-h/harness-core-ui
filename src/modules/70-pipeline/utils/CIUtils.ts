/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { get, isEmpty } from 'lodash-es'
import moment from 'moment'
import { RUNTIME_INPUT_VALUE, SelectOption } from '@harness/uicore'
import type { GitFilterScope } from '@common/components/GitFilters/GitFilters'
import { usePipelineContext } from '@pipeline/components/PipelineStudio/PipelineContext/PipelineContext'
import { RegExAllowedInputExpression } from '@pipeline/components/PipelineSteps/Steps/CustomVariables/CustomVariableInputSet'
import type { GitQueryParams } from '@common/interfaces/RouteInterfaces'
import { useQueryParams } from '@common/hooks'
import type { StepViewType } from '@pipeline/components/AbstractSteps/Step'
import { ConnectorRefWidth } from './constants'

// Returns first 7 letters of commit ID
export function getShortCommitId(commitId: string): string {
  return commitId.slice(0, 7)
}

// TODO: Add singular forms, better using i18n because they have support for it
export function getTimeAgo(timeStamp: number): string {
  const currentDate = moment(new Date())
  const timeStampAsDate = moment(timeStamp)

  if (currentDate.diff(timeStampAsDate, 'days') > 30) {
    return `on ${timeStampAsDate.format('MMM D')}`
  } else if (currentDate.diff(timeStampAsDate, 'days') === 1) {
    return 'yesterday'
  } else if (currentDate.diff(timeStampAsDate, 'days') === 0) {
    if (currentDate.diff(timeStampAsDate, 'minutes') >= 60) {
      return `${currentDate.diff(timeStampAsDate, 'hours')} hours ago`
    } else {
      return `${currentDate.diff(timeStampAsDate, 'minutes')} minutes ago`
    }
  } else {
    return `${currentDate.diff(timeStampAsDate, 'days')} days ago`
  }
}

export function useGitScope(): GitFilterScope | undefined {
  const gitDetails = usePipelineContext()?.state?.gitDetails
  const { repoIdentifier, branch } = useQueryParams<GitQueryParams>()

  if (!isEmpty(gitDetails)) {
    return {
      repo: gitDetails.repoIdentifier!,
      branch: gitDetails.branch!,
      getDefaultFromOtherRepo: true
    }
  } else if (!!repoIdentifier && !!branch) {
    return {
      repo: repoIdentifier,
      branch,
      getDefaultFromOtherRepo: true
    }
  }
}

export const getAllowedValuesFromTemplate = (template: Record<string, any>, fieldPath: string): SelectOption[] => {
  if (!template || !fieldPath) {
    return []
  }
  const value = get(template, fieldPath, '')
  const items: SelectOption[] = []
  if (RegExAllowedInputExpression.test(value as string)) {
    // This separates out "<+input>.allowedValues(a, b, c)" to ["<+input>", ["a", "b", "c"]]
    const match = (value as string).match(RegExAllowedInputExpression)
    if (match && match?.length > 1) {
      const allowedValues = match[1]
      items.push(...allowedValues.split(',').map(item => ({ label: item, value: item })))
    }
  }
  return items
}

export const shouldRenderRunTimeInputView = (value: any): boolean => {
  if (!value) {
    return false
  }
  if (typeof value === 'object') {
    return Object.keys(value).some(key => typeof value[key] === 'string' && value[key].startsWith(RUNTIME_INPUT_VALUE))
  } else {
    return typeof value === 'string' && value.startsWith(RUNTIME_INPUT_VALUE)
  }
}

export const shouldRenderRunTimeInputViewWithAllowedValues = (
  fieldPath: string,
  template?: Record<string, any>
): boolean => {
  if (!template || !fieldPath) {
    return false
  }
  const allowedValues = get(template, fieldPath, '')
  return shouldRenderRunTimeInputView(allowedValues) && RegExAllowedInputExpression.test(allowedValues)
}

export const getConnectorRefWidth = (viewType: StepViewType): number =>
  Object.entries(ConnectorRefWidth).find(key => key[0] === viewType)?.[1] || ConnectorRefWidth.DefaultView
