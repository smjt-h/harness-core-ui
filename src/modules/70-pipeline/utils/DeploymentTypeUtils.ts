/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { IconName } from '@harness/uicore'
import type { ServiceDefinition } from 'services/cd-ng'
import type { StringKeys } from 'framework/strings'

export const deploymentTypeLabel: Record<ServiceDefinition['type'], StringKeys> = {
  Kubernetes: 'kubernetesText',
  NativeHelm: 'pipeline.nativeHelm',
  Ssh: 'SSH',
  ServerlessAwsLambda: 'pipeline.serviceDeploymentTypes.serverlessAwsLambda',
  WinRm: 'pipeline.serviceDeploymentTypes.winrm'
}

export const deploymentTypeIcon: Record<string, IconName> = {
  Kubernetes: 'service-kubernetes',
  NativeHelm: 'service-helm',
  ServerlessAwsLambda: 'service-serverless-aws',
  Ssh: 'secret-ssh',
  WinRm: 'command-winrm'
}
