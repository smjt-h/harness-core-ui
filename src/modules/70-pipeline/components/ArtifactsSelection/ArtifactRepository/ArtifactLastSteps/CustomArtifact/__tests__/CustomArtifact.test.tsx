/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { act, fireEvent, render, waitFor } from '@testing-library/react'
import { MultiTypeInputType, RUNTIME_INPUT_VALUE } from '@wings-software/uicore'
import { queryByNameAttribute, TestWrapper } from '@common/utils/testUtils'
import type { ArtifactType, CustomArtifactSource } from '@pipeline/components/ArtifactsSelection/ArtifactInterface'
import { ServiceDeploymentType } from '@pipeline/utils/stageHelpers'
import { CustomArtifact } from '../CustomArtifact'

const props = {
  name: 'Artifact details',
  expressions: [],
  allowableTypes: [MultiTypeInputType.FIXED, MultiTypeInputType.RUNTIME, MultiTypeInputType.EXPRESSION],
  context: 2,
  handleSubmit: jest.fn(),
  artifactIdentifiers: [],
  selectedArtifact: 'CustomArtifact' as ArtifactType,
  selectedDeploymentType: ServiceDeploymentType.Kubernetes
}

const initialValues = {
  identifier: '',
  version: RUNTIME_INPUT_VALUE
}

describe('Nexus Artifact tests', () => {
  test(`renders without crashing`, () => {
    const { container } = render(
      <TestWrapper>
        <CustomArtifact key={'key'} initialValues={initialValues as CustomArtifactSource} {...props} />
      </TestWrapper>
    )
    expect(container).toMatchSnapshot()
  })
  test(`version should have default value of runtimeinput`, async () => {
    const { container } = render(
      <TestWrapper>
        <CustomArtifact key={'key'} initialValues={initialValues as CustomArtifactSource} {...props} />
      </TestWrapper>
    )
    const versionField = queryByNameAttribute('version', container) as HTMLInputElement
    expect(versionField).toBeTruthy()
    expect(versionField.value).toEqual(RUNTIME_INPUT_VALUE)

    const submitBtn = container.querySelector('button[type="submit"]')!
    fireEvent.click(submitBtn)
  })

  test(`able to submit form when the form is non empty`, async () => {
    const { container } = render(
      <TestWrapper>
        <CustomArtifact key={'key'} initialValues={initialValues as CustomArtifactSource} {...props} />
      </TestWrapper>
    )
    const submitBtn = container.querySelector('button[type="submit"]')!
    fireEvent.click(submitBtn)

    await act(async () => {
      fireEvent.change(queryByNameAttribute('identifier', container)!, { target: { value: 'testidentifier' } })
    })
    fireEvent.click(submitBtn)

    await waitFor(() => {
      expect(props.handleSubmit).toBeCalled()
      expect(props.handleSubmit).toHaveBeenCalledWith({
        identifier: 'testidentifier',
        spec: {
          version: '<+input>'
        }
      })
    })
  })

  test(`form renders correctly in Edit Case`, async () => {
    const filledInValues = {
      identifier: 'nexusSidecarId',
      version: 'artifact-version'
    }
    const { container } = render(
      <TestWrapper>
        <CustomArtifact key={'key'} initialValues={filledInValues as CustomArtifactSource} {...props} />
      </TestWrapper>
    )
    const versionField = container.querySelector('input[name="version"]')
    expect(versionField).not.toBeNull()

    expect(container).toMatchSnapshot()
  })
})
