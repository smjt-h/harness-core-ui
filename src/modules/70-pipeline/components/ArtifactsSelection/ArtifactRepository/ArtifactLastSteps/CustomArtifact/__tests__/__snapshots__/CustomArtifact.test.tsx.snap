// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nexus Artifact tests form renders correctly in Edit Case 1`] = `
<div>
  <div
    class="Layout--vertical Layout--layout-spacing-medium StyledProps--main firstep"
  >
    <p
      class="StyledProps--font StyledProps--main StyledProps--font-variation-h3 StyledProps--margin-bottom-medium"
    >
      pipeline.artifactsSelection.artifactDetails
    </p>
    <div
      class="OverlaySpinner--overlaySpinner"
    >
      <form
        action="#"
      >
        <div
          class="connectorForm"
        >
          <div
            class="dockerSideCard"
          >
            <div
              class="bp3-form-group"
            >
              <label
                class="bp3-label"
                for="identifier"
              >
                <span
                  class="Tooltip--acenter"
                  data-tooltip-id="imagePath_identifier"
                >
                  pipeline.artifactsSelection.existingDocker.sidecarId
                </span>
                 
                <span
                  class="bp3-text-muted"
                />
              </label>
              <div
                class="bp3-form-content"
              >
                <div
                  class="bp3-input-group"
                >
                  <input
                    autocomplete="off"
                    class="bp3-input"
                    name="identifier"
                    placeholder="pipeline.artifactsSelection.existingDocker.sidecarIdPlaceholder"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="imagePathContainer"
          >
            <div
              class="bp3-form-group"
            >
              <label
                class="bp3-label"
                for="version"
              >
                <span
                  class="Tooltip--acenter"
                  data-tooltip-id="imagePath_version"
                >
                  version
                </span>
                 
                <span
                  class="bp3-text-muted"
                />
              </label>
              <div
                class="bp3-form-content"
              >
                <div
                  class="Layout--horizontal StyledProps--main MultiTypeInput--main MultiTypeInput--disabled"
                >
                  <div
                    class="TextInput--main MultiTypeInput--input"
                  >
                    <div
                      class="bp3-input-group bp3-disabled"
                    >
                      <input
                        class="bp3-input"
                        disabled=""
                        name="version"
                        placeholder="<+input>"
                        type="text"
                        value="<+input>"
                      />
                    </div>
                  </div>
                  <span
                    class="bp3-popover-wrapper MultiTypeInput--wrapper"
                  >
                    <span
                      class="bp3-popover-target"
                    >
                      <button
                        class="MultiTypeInput--btn MultiTypeInput--RUNTIME"
                      >
                        <span
                          class="bp3-icon StyledProps--main"
                          data-icon="runtime-input"
                        >
                          <svg
                            height="12"
                            viewBox="0 0 10 7.97"
                            width="12"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M8.14 4L8 3.72a2.53 2.53 0 01-.12-.3l-.16-.35c-.11-.25-.22-.38-.32-.38s-.28.15-.37.44C7 3.1 7 3.07 7 3a.73.73 0 01.24-.46.66.66 0 01.42-.26c.25 0 .5.35.76 1.05l.07.17.07-.12c.47-.73.86-1.1 1.18-1.1a.71.71 0 01.29.07l-.4.42h-.11c-.24 0-.52.28-.86.83l-.09.16.07.2c.29.79.53 1.18.71 1.18s.31-.12.4-.36q.09.06.09.12t-.27.39a.74.74 0 01-.45.23c-.25 0-.51-.36-.77-1.07l-.12-.25-.1.17c-.44.76-.87 1.14-1.28 1.14a.65.65 0 01-.39-.12L6.84 5a.32.32 0 00.23.1c.25 0 .53-.25.83-.74l.17-.28zm-4-.32L1.7.59h2.47S5 .52 5.35 1a3.28 3.28 0 01.42 1.14h.31L5.94 0H0v.31l3.06 3.74L0 7.62V8h6.08l.38-2.36-.31-.07a2.43 2.43 0 01-.52 1c-.42.48-2 .41-2 .41H1.39z"
                              fill="current-color"
                            />
                          </svg>
                        </span>
                      </button>
                    </span>
                  </span>
                </div>
              </div>
            </div>
            <div
              class="configureOptions"
            >
              <span
                class="bp3-popover-wrapper bp3-dark"
              >
                <span
                  class="bp3-popover-target"
                >
                  <button
                    class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main Button--with-current-color Button--iconOnly Button--without-shadow Button--withRightIcon"
                    id="configureOptions_version"
                    style="align-self: center;"
                    tabindex="0"
                    type="button"
                  >
                    <span
                      class="bp3-icon bp3-icon-cog StyledProps--main"
                      icon="cog"
                    >
                      <svg
                        data-icon="cog"
                        height="16"
                        viewBox="0 0 16 16"
                        width="16"
                      >
                        <desc>
                          cog
                        </desc>
                        <path
                          d="M15.19 6.39h-1.85c-.11-.37-.27-.71-.45-1.04l1.36-1.36c.31-.31.31-.82 0-1.13l-1.13-1.13a.803.803 0 00-1.13 0l-1.36 1.36c-.33-.17-.67-.33-1.04-.44V.79c0-.44-.36-.8-.8-.8h-1.6c-.44 0-.8.36-.8.8v1.86c-.39.12-.75.28-1.1.47l-1.3-1.3c-.3-.3-.79-.3-1.09 0L1.82 2.91c-.3.3-.3.79 0 1.09l1.3 1.3c-.2.34-.36.7-.48 1.09H.79c-.44 0-.8.36-.8.8v1.6c0 .44.36.8.8.8h1.85c.11.37.27.71.45 1.04l-1.36 1.36c-.31.31-.31.82 0 1.13l1.13 1.13c.31.31.82.31 1.13 0l1.36-1.36c.33.18.67.33 1.04.44v1.86c0 .44.36.8.8.8h1.6c.44 0 .8-.36.8-.8v-1.86c.39-.12.75-.28 1.1-.47l1.3 1.3c.3.3.79.3 1.09 0l1.09-1.09c.3-.3.3-.79 0-1.09l-1.3-1.3c.19-.35.36-.71.48-1.1h1.85c.44 0 .8-.36.8-.8v-1.6a.816.816 0 00-.81-.79zm-7.2 4.6c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3z"
                          fill-rule="evenodd"
                        />
                      </svg>
                    </span>
                  </button>
                </span>
              </span>
            </div>
          </div>
        </div>
        <div
          class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main"
        >
          <button
            class="bp3-button Button--button StyledProps--font StyledProps--main Button--with-current-color Button--withLeftIcon Button--variation Button--variation-secondary"
            type="button"
          >
            <span
              class="bp3-icon bp3-icon-chevron-left StyledProps--main StyledProps--padding-right-xsmall"
              icon="chevron-left"
            >
              <svg
                data-icon="chevron-left"
                height="16"
                viewBox="0 0 16 16"
                width="16"
              >
                <desc>
                  chevron-left
                </desc>
                <path
                  d="M7.41 8l3.29-3.29c.19-.18.3-.43.3-.71a1.003 1.003 0 00-1.71-.71l-4 4C5.11 7.47 5 7.72 5 8c0 .28.11.53.29.71l4 4a1.003 1.003 0 001.42-1.42L7.41 8z"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
            <span
              class="bp3-button-text"
            >
              back
            </span>
          </button>
          <button
            class="bp3-button Button--button StyledProps--font StyledProps--main Button--with-current-color Button--withRightIcon Button--variation Button--variation-primary"
            type="submit"
          >
            <span
              class="bp3-button-text"
            >
              submit
            </span>
            <span
              class="bp3-icon bp3-icon-chevron-right StyledProps--main"
              icon="chevron-right"
            >
              <svg
                data-icon="chevron-right"
                height="16"
                viewBox="0 0 16 16"
                width="16"
              >
                <desc>
                  chevron-right
                </desc>
                <path
                  d="M10.71 7.29l-4-4a1.003 1.003 0 00-1.42 1.42L8.59 8 5.3 11.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71l4-4c.18-.18.29-.43.29-.71 0-.28-.11-.53-.29-.71z"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`Nexus Artifact tests renders without crashing 1`] = `
<div>
  <div
    class="Layout--vertical Layout--layout-spacing-medium StyledProps--main firstep"
  >
    <p
      class="StyledProps--font StyledProps--main StyledProps--font-variation-h3 StyledProps--margin-bottom-medium"
    >
      pipeline.artifactsSelection.artifactDetails
    </p>
    <div
      class="OverlaySpinner--overlaySpinner"
    >
      <form
        action="#"
      >
        <div
          class="connectorForm"
        >
          <div
            class="dockerSideCard"
          >
            <div
              class="bp3-form-group"
            >
              <label
                class="bp3-label"
                for="identifier"
              >
                <span
                  class="Tooltip--acenter"
                  data-tooltip-id="imagePath_identifier"
                >
                  pipeline.artifactsSelection.existingDocker.sidecarId
                </span>
                 
                <span
                  class="bp3-text-muted"
                />
              </label>
              <div
                class="bp3-form-content"
              >
                <div
                  class="bp3-input-group"
                >
                  <input
                    autocomplete="off"
                    class="bp3-input"
                    name="identifier"
                    placeholder="pipeline.artifactsSelection.existingDocker.sidecarIdPlaceholder"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="imagePathContainer"
          >
            <div
              class="bp3-form-group"
            >
              <label
                class="bp3-label"
                for="version"
              >
                <span
                  class="Tooltip--acenter"
                  data-tooltip-id="imagePath_version"
                >
                  version
                </span>
                 
                <span
                  class="bp3-text-muted"
                />
              </label>
              <div
                class="bp3-form-content"
              >
                <div
                  class="Layout--horizontal StyledProps--main MultiTypeInput--main MultiTypeInput--disabled"
                >
                  <div
                    class="TextInput--main MultiTypeInput--input"
                  >
                    <div
                      class="bp3-input-group bp3-disabled"
                    >
                      <input
                        class="bp3-input"
                        disabled=""
                        name="version"
                        placeholder="<+input>"
                        type="text"
                        value="<+input>"
                      />
                    </div>
                  </div>
                  <span
                    class="bp3-popover-wrapper MultiTypeInput--wrapper"
                  >
                    <span
                      class="bp3-popover-target"
                    >
                      <button
                        class="MultiTypeInput--btn MultiTypeInput--RUNTIME"
                      >
                        <span
                          class="bp3-icon StyledProps--main"
                          data-icon="runtime-input"
                        >
                          <svg
                            height="12"
                            viewBox="0 0 10 7.97"
                            width="12"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M8.14 4L8 3.72a2.53 2.53 0 01-.12-.3l-.16-.35c-.11-.25-.22-.38-.32-.38s-.28.15-.37.44C7 3.1 7 3.07 7 3a.73.73 0 01.24-.46.66.66 0 01.42-.26c.25 0 .5.35.76 1.05l.07.17.07-.12c.47-.73.86-1.1 1.18-1.1a.71.71 0 01.29.07l-.4.42h-.11c-.24 0-.52.28-.86.83l-.09.16.07.2c.29.79.53 1.18.71 1.18s.31-.12.4-.36q.09.06.09.12t-.27.39a.74.74 0 01-.45.23c-.25 0-.51-.36-.77-1.07l-.12-.25-.1.17c-.44.76-.87 1.14-1.28 1.14a.65.65 0 01-.39-.12L6.84 5a.32.32 0 00.23.1c.25 0 .53-.25.83-.74l.17-.28zm-4-.32L1.7.59h2.47S5 .52 5.35 1a3.28 3.28 0 01.42 1.14h.31L5.94 0H0v.31l3.06 3.74L0 7.62V8h6.08l.38-2.36-.31-.07a2.43 2.43 0 01-.52 1c-.42.48-2 .41-2 .41H1.39z"
                              fill="current-color"
                            />
                          </svg>
                        </span>
                      </button>
                    </span>
                  </span>
                </div>
              </div>
            </div>
            <div
              class="configureOptions"
            >
              <span
                class="bp3-popover-wrapper bp3-dark"
              >
                <span
                  class="bp3-popover-target"
                >
                  <button
                    class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main Button--with-current-color Button--iconOnly Button--without-shadow Button--withRightIcon"
                    id="configureOptions_version"
                    style="align-self: center;"
                    tabindex="0"
                    type="button"
                  >
                    <span
                      class="bp3-icon bp3-icon-cog StyledProps--main"
                      icon="cog"
                    >
                      <svg
                        data-icon="cog"
                        height="16"
                        viewBox="0 0 16 16"
                        width="16"
                      >
                        <desc>
                          cog
                        </desc>
                        <path
                          d="M15.19 6.39h-1.85c-.11-.37-.27-.71-.45-1.04l1.36-1.36c.31-.31.31-.82 0-1.13l-1.13-1.13a.803.803 0 00-1.13 0l-1.36 1.36c-.33-.17-.67-.33-1.04-.44V.79c0-.44-.36-.8-.8-.8h-1.6c-.44 0-.8.36-.8.8v1.86c-.39.12-.75.28-1.1.47l-1.3-1.3c-.3-.3-.79-.3-1.09 0L1.82 2.91c-.3.3-.3.79 0 1.09l1.3 1.3c-.2.34-.36.7-.48 1.09H.79c-.44 0-.8.36-.8.8v1.6c0 .44.36.8.8.8h1.85c.11.37.27.71.45 1.04l-1.36 1.36c-.31.31-.31.82 0 1.13l1.13 1.13c.31.31.82.31 1.13 0l1.36-1.36c.33.18.67.33 1.04.44v1.86c0 .44.36.8.8.8h1.6c.44 0 .8-.36.8-.8v-1.86c.39-.12.75-.28 1.1-.47l1.3 1.3c.3.3.79.3 1.09 0l1.09-1.09c.3-.3.3-.79 0-1.09l-1.3-1.3c.19-.35.36-.71.48-1.1h1.85c.44 0 .8-.36.8-.8v-1.6a.816.816 0 00-.81-.79zm-7.2 4.6c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3z"
                          fill-rule="evenodd"
                        />
                      </svg>
                    </span>
                  </button>
                </span>
              </span>
            </div>
          </div>
        </div>
        <div
          class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main"
        >
          <button
            class="bp3-button Button--button StyledProps--font StyledProps--main Button--with-current-color Button--withLeftIcon Button--variation Button--variation-secondary"
            type="button"
          >
            <span
              class="bp3-icon bp3-icon-chevron-left StyledProps--main StyledProps--padding-right-xsmall"
              icon="chevron-left"
            >
              <svg
                data-icon="chevron-left"
                height="16"
                viewBox="0 0 16 16"
                width="16"
              >
                <desc>
                  chevron-left
                </desc>
                <path
                  d="M7.41 8l3.29-3.29c.19-.18.3-.43.3-.71a1.003 1.003 0 00-1.71-.71l-4 4C5.11 7.47 5 7.72 5 8c0 .28.11.53.29.71l4 4a1.003 1.003 0 001.42-1.42L7.41 8z"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
            <span
              class="bp3-button-text"
            >
              back
            </span>
          </button>
          <button
            class="bp3-button Button--button StyledProps--font StyledProps--main Button--with-current-color Button--withRightIcon Button--variation Button--variation-primary"
            type="submit"
          >
            <span
              class="bp3-button-text"
            >
              submit
            </span>
            <span
              class="bp3-icon bp3-icon-chevron-right StyledProps--main"
              icon="chevron-right"
            >
              <svg
                data-icon="chevron-right"
                height="16"
                viewBox="0 0 16 16"
                width="16"
              >
                <desc>
                  chevron-right
                </desc>
                <path
                  d="M10.71 7.29l-4-4a1.003 1.003 0 00-1.42 1.42L8.59 8 5.3 11.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71l4-4c.18-.18.29-.43.29-.71 0-.28-.11-.53-.29-.71z"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
